gateway/ln-gateway/src/bin/cln_extension.rs

```diff
@@ -0,0 +1,330 @@
+            .call(cln_rpc::Request::Getinfo(
+                model::requests::GetinfoRequest {},
+            ))
+            .await
+            .map(|response| {
+                let pub_key = match response {
+                    cln_rpc::Response::Getinfo(model::responses::GetinfoResponse {
+                        id, ..
+                    }) => id,
+                    _ => panic!("Unexpected response from cln_rpc"),
+                };
+                tonic::Response::new(GetPubKeyResponse {
+                    pub_key: pub_key.serialize().to_vec(),
+                })
+            })
+            .map_err(|e| {
+                error!("cln getinfo returned error: {:?}", e);
+                Status::internal(e.to_string())
+            })
+    }
+
+    async fn pay_invoice(
+        &self,
+        request: tonic::Request<PayInvoiceRequest>,
+    ) -> Result<tonic::Response<PayInvoiceResponse>, Status> {
+        let PayInvoiceRequest {
+            invoice,
+            max_fee_percent,
+            max_delay,
+        } = request.into_inner();
+
+        self.client
+            .lock()
+            .await
+            .call(cln_rpc::Request::Pay(model::PayRequest {
+                bolt11: invoice,
+                amount_msat: None,
+                label: None,
+                riskfactor: None,
+                maxfeepercent: Some(max_fee_percent),
+                retry_for: None,
+                maxdelay: Some(max_delay as u16),
+                exemptfee: None,
+                localinvreqid: None,
+                exclude: None,
+                maxfee: None,
+                description: None,
+            }))
+            .await
+            .map(|response| {
+                let pay_success = match response {
+                    cln_rpc::Response::Pay(model::PayResponse {
+                        payment_preimage, ..
+                    }) => PayInvoiceResponse {
+                        preimage: payment_preimage.to_vec(),
+                    },
+                    _ => panic!("Unexpected response from cln_rpc"),
+                };
+                tonic::Response::new(pay_success)
+            })
+            .map_err(|e| {
+                error!("cln pay returned error {:?}", e);
+                Status::internal(e.to_string())
+            })
+    }
+
+    type SubscribeInterceptHtlcsStream =
+        ReceiverStream<Result<SubscribeInterceptHtlcsResponse, Status>>;
+
+    async fn subscribe_intercept_htlcs(
+        &self,
+        request: tonic::Request<SubscribeInterceptHtlcsRequest>,
+    ) -> Result<tonic::Response<Self::SubscribeInterceptHtlcsStream>, Status> {
+        let SubscribeInterceptHtlcsRequest { short_channel_id } = request.into_inner();
+        let receiver = self
+            .interceptor
+            .lock()
+            .await
+            .add_htlc_subscriber(short_channel_id)
+            .await;
+
+        Ok(tonic::Response::new(ReceiverStream::new(receiver)))
+    }
```

Comment:
Does this create a persistent connection? Anything need to happen if that connection is dropped?

Reply:
Client will detect the disconnect and can just open a new subscription. In the meanwhile, the extension service will :
- continue intercepting htlcs (memory of a previous subscription as indicated by `short_channel_id` in subscriptions map)
- attempt to send intercepted htlc to a broken / stale subscription. FAIL
- wait out the htlc and automatically resolve it after timeout

If a client opens a new subscription with same `short_channel_id`, it refreshes / replaces the previous sub and now the extension can send new intercepted htlcs for processing

I'm yet to test out this logic. Any improvements possible?